File: julia/examples/simple_test.jl
# Construct a simple test problem using JuMP. The problem is written as:
# min   x'*x
# s.t.  x[1] = 1
#       x[2] ≥ 1
#       0 ≤ (x[3] + 1) ⟂ (x[4] - 1) ≥ 0
# The solution is x = [1, 1, 0, 1] with an objective of 3
using Pkg; Pkg.activate(@__DIR__)
using Revise
using JuMP, Marble, NLPModelsJuMP

# Construct problem using JuMP
model = JuMP.Model()

@variable(model, x[1:4])
@objective(model, Min, x'*x)
@constraint(model, x[1] == 1)
@constraint(model, x[2] >= 1)

# For complementarities, we define each inequality and then specify indices
# for the complementarity pairs
@constraint(model, x3_comp, x[3] + 1 >= 0)
@constraint(model, x4_comp, x[4] - 1 >= 0)

mpcc = MPCC(model,
    (model[:x3_comp], model[:x4_comp])
)

solver = Marble.Solver()
Marble.setup!(solver, mpcc; verbosity = 1)
results = Marble.solve!(solver)
z = Marble.z(results)

println(z)
println(Marble.obj(solver, z))
println(Marble.residual_eq(solver, z))
println(Marble.residual_ineq(solver, z))
println(Marble.residual_comp(solver, z))

File: python/examples/simple_test.py
# Construct a simple test problem minimizing x'x
# where x[1] = 1
#       x[2] >= 1
#       0 <= (x[3] + 1) ⟂ (x[4] - 1) >= 0  -->  solution is x[3] = 0, x[4] = 1
import numpy as np
import marble

Q = 2.0 * np.eye(4)
q = np.zeros(4)

J_eq,   c_eq   = [[1.0, 0.0, 0.0, 0.0]], [-1.0]   # x[1] == 1
J_ineq, c_ineq = [[0.0, 1.0, 0.0, 0.0]], [-1.0]   # x[2] >= 1
L, l           = [[0.0, 0.0, 1.0, 0.0]], [1.0]    # x[3] + 1
R, r           = [[0.0, 0.0, 0.0, 1.0]], [-1.0]   # x[4] - 1

solver = marble.Solver()
solver.setup(Q, q, 0.0,
             J_eq=J_eq, c_eq=c_eq, J_ineq=J_ineq, c_ineq=c_ineq,
             L=L, l=l, R=R, r=r, verbosity=1)
results = solver.solve()

z = np.asarray(results.z)
print(z)
print(solver.problem.obj(z))
print(solver.problem.residual_eq(z))
print(solver.problem.residual_ineq(z))
print(solver.problem.residual_comp(z))
